src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+//! ALBA's bounded DFS scheme using Blake2b as hash function.
+//! (c.f. Section 3.2.2 of Alba paper)
+
+use rayon::prelude::*;
+use std::sync::atomic::{self, AtomicUsize};
+
+extern crate core;
+use crate::utils;
+
+use std::f64::consts::E;
+use std::sync::Arc;
+
+const DATA_LENGTH: usize = 32;
+const DIGEST_SIZE: usize = 32;
+
+type Data = [u8; DATA_LENGTH];
+type Hash = [u8; DIGEST_SIZE];
+
+/// Setup input parameters
+#[derive(Debug, Clone)]
+pub struct Params {
+    /// Soundness security parameter
+    pub lambda_sec: usize,
+    /// Completeness security parameter
+    pub lambda_rel: usize,
+    /// Approximate size of set Sp to lower bound
+    pub n_p: usize,
+    /// Target lower bound
+    pub n_f: usize,
+}
+pub enum Cases {
+    /// Case where u =< λ^2
+    Small,
+    /// Case where λ^2 < u < λ^3
+    Mid,
+    /// Case where u >= λ^3
+    High,
+}
+
+impl Params {
+    /// Returns information on which case corresponds some parameter
+    pub fn which_case(&self) -> (Cases, usize) {
+        let lsec = self.lambda_sec as f64;
+        let lrel = self.lambda_rel as f64;
+        let np = self.n_p as f64;
+        let nf = self.n_f as f64;
+        let loge = E.log2();
+
+        let lognpnf = (np / nf).log2();
+        let u_f64 = (lsec + lrel.log2() + 5.0 - loge.log2()) / lognpnf;
+        let u = u_f64.ceil() as u64;
+
+        let ratio = 9.0 * np * loge / ((17 * u).pow(2) as f64);
+        let s1 = ratio - 7.0;
+        let s2 = ratio - 2.0;
+
+        if s1 < 1.0 || s2 < 1.0 {
+            return (Cases::Small, u as usize);
+        }
+
+        let lrel2 = lrel.min(s2);
+        if (u as f64) < lrel2 {
+            (Cases::Mid, u as usize)
+        } else {
+            (Cases::High, u as usize)
+        }
+    }
+}
+
+/// Setup output parameters
+#[derive(Debug, Clone)]
+pub struct Setup {
+    /// Approximate size of set Sp to lower bound
+    pub n_p: usize,
+    /// Proof size (in Sp elements)
+    pub u: usize,
+    /// Proof max counter
+    pub r: usize,
+    /// Proof max 2nd counter
+    pub d: usize,
+    /// Probability q
+    pub q: f64,
+    /// Computation bound
+    pub b: usize,
+}
+impl Setup {
+    /// Setup algorithm taking a Params as input and returning setup parameters (u,d,q)
+    pub fn new(params: &Params) -> Self {
```

Comment:
I'd think about ways to refactor this constructor to make the code a bit more readable and also, avoid code duplications. Here are a few ideas: 

1) extract a function to compute the u parameter outside of the constructor and simply call this function 

2) extract a function (e.g., `foo()`)  to determine which case we are into (Theorem 13, 14, etc.). This function can give you back also the appropriate parameters for each case. 

3) Use a match expression to construct the different setup parameters. Something like 

```
match foo() {
  Case::Small { <params for small> } {
     // construct the Setup object for the small case 
  }
  // other cases
}
```